aopy/postproc/centerout.py

```diff
@@ -0,0 +1,237 @@
+    new = np.linalg.multi_dot((old, offset_arr, scale_arr, rotations[rotation], exp_rotations[exp_rotation]))
+    pertubation_rot = R.from_euler('y', perturbation, degrees=True)
+    hand_transformed = np.matmul(pertubation_rot.as_matrix(), new[0:3])
+#     print(coords, new_coords)
+    return hand_transformed
+
+def get_taskspace_and_nullspace(mapping, task='2DCenterOut'):
+    '''
+    Decomposes hand movement into components of movement in the plane of the screen and out of the plane of the screen
+    Args:
+        mapping (3 x 3 numpy array): Mapping matrix used in experiment. See :func:`get_mapping` for more details
+
+    Returns:
+        t_a (3 x 3 numpy array): Matrix that projects hand movement into the task potent cursor space
+        n_a (3 x 3 numpy array): Matrix that projects hand movement into the null space
+    '''
+    if task == '2DCenterOut':
+        a = mapping[[0,2], :] # rows are cursor and columns are hand. taking only hy and hz components of optitrack that correspond to up/down and right/left
+    elif task == '1DTracking':
+        a = mapping[2,:]
+
+    a_plus = a.T @ np.linalg.pinv(a @ a.T)
+    t_a = a_plus @ a
+    n_a = np.eye(3) - t_a
+
+    return t_a, n_a
+
+def decompose_hand_movements(hand_data, mapping, task='2DCenterOut'):
+    '''
+    Decomposes hand movement into components of movement in the plane of the screen and out of the plane of the screen
+    Args:
+        hand_data (2D numpy array): 3D array of hand trajectory per trial (n_timepoints x 3) in optitrack space. X - forward/back, Y - up/down, Z - right/left
+        mapping (3 x 3 numpy array): Mapping matrix used in experiment. See :func:`_get_mapping` for more details . Rows must be cursor and columns must be hand.
+
+    Returns:
+
+    '''
+    t_a, n_a = get_taskspace_and_nullspace(mapping, task)
+    hand_data = np.array(hand_data)
+
+    hT = np.dot(t_a, hand_data.T)  # 3 x n_timepoints
+    hN = np.dot(n_a, hand_data.T)
+
+    hT_norm = np.mean(np.linalg.norm(hT, axis=0)) # gives the magnitude of movement in task space
+    hN_norm = np.mean(np.linalg.norm(hN, axis=0)) # gives the magnitude of movement in null space
+
+    return  hT_norm, hN_norm
+
+
+def transform_optitrack2hand_coordinates(o_coords):
+    """
+    Transforms coordinates from the Optitrack coordinates (O) to the intuitive hand coordinates for plotting (H).
+
+    Parameters:
+        o_coords (numpy array): Optitrack data in Optitrack coordinates [Ox, Oy, Oz] as a numpy array (n_timepoints x 3).
+
+    Returns:
+        numpy array: The transformed coordinates [Hx, Hy, Hz].
+
+    .. image:: _images/hand_data_coordinates.png
+
+    """
+    # Transformation matrix
+    T = np.array([
+        [0, 0, 1],
+        [0, 1, 0],
+        [1, 0, 0]
+    ])
+
+    # Perform the matrix multiplication
+    h_coords = o_coords.dot(T)
+
+    return h_coords
```

Comment:
shouldn't this also include the offset and scale portion of the full `_transform_coords()`?